packages/jest-runtime/src/transform.js

```diff
@@ -110,94 +116,189 @@ const readCacheFile = (filePath: Path, cachePath: Path): ?string => {
-    const cacheDir = path.join(baseCacheDir, cacheKey[0] + cacheKey[1]);
-    const cachePath = path.join(
-      cacheDir,
-      path.basename(filename, path.extname(filename)) + '_' + cacheKey,
-    );
-    createDirectory(cacheDir);
+  if (config.coveragePathIgnorePatterns &&
+    config.coveragePathIgnorePatterns.some(pattern => filename.match(pattern))
+  ) {
+    return false;
+  }
+
+  if (filename.match(config.mocksPattern)) {
+    return false;
+  }
+
+  return true;
+};
 
-    const cacheData = config.cache ? readCacheFile(filename, cachePath) : null;
-    if (cacheData) {
-      content = cacheData;
+const getFileCachePath = (
+  filename: Path,
+  config: Config,
+  content: string,
+): Path => {
+  const baseCacheDir = getCacheFilePath(
+    config.cacheDirectory,
+    'jest-transform-cache-' + config.name,
+    VERSION,
+  );
+  const cacheKey = getCacheKey(content, filename, config);
+  // Create sub folders based on the cacheKey to avoid creating one
+  // directory with many files.
+  const cacheDir = path.join(baseCacheDir, cacheKey[0] + cacheKey[1]);
+  const cachePath = path.join(
+    cacheDir,
+    path.basename(filename, path.extname(filename)) + '_' + cacheKey,
+  );
+  createDirectory(cacheDir);
+
+  return cachePath;
+};
+
+const preprocessorCache = new WeakMap();
+
+const getPreprocessor = (config: Config): ?Processor => {
+  if (preprocessorCache.has(config)) {
+    return preprocessorCache.get(config);
+  } else {
+    let preprocessor;
+    if (!config.scriptPreprocessor) {
+      preprocessor = null;
     } else {
-      content = preprocessor.process(content, filename, config);
-      if (options && options.instrument) {
-        content = options.instrument(content);
+      // $FlowFixMe
+      preprocessor = require(config.scriptPreprocessor);
+      if (typeof preprocessor.process !== 'function') {
+        throw new TypeError(
+          'Jest: a preprocessor must export a `process` function.',
+        );
       }
-      content = wrap(content);
-      writeCacheFile(cachePath, content);
     }
-  } else if (options && options.instrument) {
-    content = wrap(options.instrument(content));
+    preprocessorCache.set(config, preprocessor);
+    return preprocessor;
+  }
+};
+
+const stripShebang = content => {
+  // If the file data starts with a shebang remove it. Leaves the empty line
+  // to keep stack trace line numbers correct.
+  if (content.startsWith('#!')) {
+    return content.replace(/^#!.*/, '');
+  } else {
+    return content;
+  }
+};
+
+const instrument = (content: string, filename: Path): string => {
+  // NOTE: Keeping these requires inside this function reduces a single run
+  // time by 2sec if not running in `--coverage` mode
+  const babel = require('babel-core');
+  const babelPluginIstanbul = require('babel-plugin-istanbul').default;
+
+  if (babel.util.canCompile(filename)) {
+    return babel.transform(content, {
+      filename,
+      plugins: [babelPluginIstanbul],
```

Comment:
Copy this line: https://github.com/facebook/jest/blob/master/packages/babel-jest/src/index.js#L18 to this transform here as well (for people not using babel-jest).

Not sure if that would duplicate this, but once we add the flag this code here isn't run.